[GTK][WPE] Fix relocation issues running the MiniBrowser or the layout tests when the path on disk changes

clopez opened the pull request:
#### 79f5dfafe126501c435ea8a87fe4dc725b62e1df
<pre>
[GTK][WPE] Fix relocation issues running the MiniBrowser or the layout tests when the path on disk changes
<a href="https://bugs.webkit.org/show_bug.cgi?id=247985">https://bugs.webkit.org/show_bug.cgi?id=247985</a>

Reviewed by Žan Doberšek.

When WebKit is built the path of the repository (where you have the checkout)
is stored in the binaries in different ways:

 - The path to the WebKit libraries is stored as an rpath on the ELF binaries
 - The CMake system also defines some constant strings to the build like WEBKIT_EXEC_PATH
   and WEBKIT_INJECTED_BUNDLE_PATH that end going into the binaries as default values.

This default values can be overriden at run-time via environment variables,
but the current tooling was not doing that.

This means that if you build WebKit in a directory and then move the directory
to a different path (without rebuilding) you are going to find errors and crashes.

This is not an issue with the flatpak build, because the flatpak build maps the
buildirectory to /app so the path doesn&apos;t really change even if you move the dir.

But this is an issue for builds not using flatpak.
Specifically I found this on a bot that I&apos;m working on, where the builder builds
on a different path than the tester (and the tester downloads the build product).

Fix the problem by passing at runtime the required paths via environment variables.

* Tools/Scripts/webkitpy/port/glib.py:
(GLibPort._prepend_to_env_value):
(GLibPort.setup_environ_for_server):
(GLibPort.setup_environ_for_minibrowser):
* Tools/Scripts/webkitpy/port/gtk.py:
(GtkPort.run_minibrowser):
* Tools/Scripts/webkitpy/port/wpe.py:
(WPEPort.setup_environ_for_minibrowser):
(WPEPort.run_minibrowser):
(WPEPort.browser_env): Deleted.
* Tools/WebKitTestRunner/InjectedBundle/gtk/ActivateFontsGtk.cpp:
(WTR::initializeFontConfigSetting):
* Tools/WebKitTestRunner/InjectedBundle/wpe/ActivateFontsWPE.cpp:
(WTR::activateFonts):
* Tools/Scripts/webkitpy/port/gtk_unittest.py

Canonical link: <a href="https://commits.webkit.org/256795@main">https://commits.webkit.org/256795@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/a760809a5148ac46e9dde87cc123489151877834

| Misc | iOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/6/builds/96836 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/77/builds/6103 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/43/builds/29933 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/8/builds/106362 "Built successfully") | [✅ 🛠 🧪 win](https://ews-build.webkit.org/#/builders/10/builds/166643 "Built successfully and passed tests") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/11/builds/100814 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/76/builds/6318 "Built successfully") | [✅ 🛠 mac-debug](https://ews-build.webkit.org/#/builders/71/builds/34831 "Built successfully") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/36/builds/89218 "Built successfully") | [✅ 🛠 wincairo](https://ews-build.webkit.org/#/builders/12/builds/103062 "Built successfully") 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/19/builds/102509 "Passed tests") | [  ~~🧪 ios-wk2~~](https://ews-build.webkit.org/#/builders/78/builds/4731 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/61/builds/83449 "Built successfully") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/35/builds/31693 "Passed tests") | 
| [✅ 🧪 webkitpy](https://ews-build.webkit.org/#/builders/5/builds/100428 "Passed tests") | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/9/builds/86567 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/3/builds/88436 "Passed tests") | [  ~~🧪 api-gtk~~](https://ews-build.webkit.org/#/builders/34/builds/74644 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/81/builds/136 "Built successfully") | [  ~~🧪 mac-wk1~~](https://ews-build.webkit.org/#/builders/73/builds/19919 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/82/builds/124 "Built successfully") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/70/builds/21339 "Passed tests") | | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/74/builds/4714 "Built successfully and passed tests") | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/80/builds/4917 "Built successfully") | [  ~~🧪 mac-AS-debug-wk2~~](https://ews-build.webkit.org/#/builders/60/builds/43866 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/79/builds/1352 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/62/builds/40638 "Passed tests") | | 
<!--EWS-Status-Bubble-End-->

zdobersek commented on Tools/Scripts/webkitpy/port/glib.py:
```diff
@@ -129,6 +136,13 @@ def setup_environ_for_server(self, server_name=None):
+        env = os.environ.copy()
+        env['WEBKIT_EXEC_PATH'] = self._build_path('bin')
+        env['WEBKIT_INJECTED_BUNDLE_PATH'] = self._build_path('lib')
+        env['LD_LIBRARY_PATH'] = self._get_path_value_updated_for_environ(env, 'LD_LIBRARY_PATH', self._build_path('lib'))
```
```
env['LD_LIBRARY_PATH'] = self._prepend_to_env_value(self._build_path('lib'), env.get('LD_LIBRAY_PATH', ''))
```
Or as you wish to name it, and the logic in the helper function can get slightly simpler.

clopez replied:
Nice! will use it :)

clopez committed:
[GTK][WPE] Fix relocation issues running the MiniBrowser or the layout tests when the path on disk changes
https://bugs.webkit.org/show_bug.cgi?id=247985

Reviewed by Žan Doberšek.

When WebKit is built the path of the repository (where you have the checkout)
is stored in the binaries in different ways:

 - The path to the WebKit libraries is stored as an rpath on the ELF binaries
 - The CMake system also defines some constant strings to the build like WEBKIT_EXEC_PATH
   and WEBKIT_INJECTED_BUNDLE_PATH that end going into the binaries as default values.

This default values can be overriden at run-time via environment variables,
but the current tooling was not doing that.

This means that if you build WebKit in a directory and then move the directory
to a different path (without rebuilding) you are going to find errors and crashes.

This is not an issue with the flatpak build, because the flatpak build maps the
buildirectory to /app so the path doesn't really change even if you move the dir.

But this is an issue for builds not using flatpak.
Specifically I found this on a bot that I'm working on, where the builder builds
on a different path than the tester (and the tester downloads the build product).

Fix the problem by passing at runtime the required paths via environment variables.

* Tools/Scripts/webkitpy/port/glib.py:
(GLibPort._prepend_to_env_value):
(GLibPort.setup_environ_for_server):
(GLibPort.setup_environ_for_minibrowser):
* Tools/Scripts/webkitpy/port/gtk.py:
(GtkPort.run_minibrowser):
* Tools/Scripts/webkitpy/port/wpe.py:
(WPEPort.setup_environ_for_minibrowser):
(WPEPort.run_minibrowser):
(WPEPort.browser_env): Deleted.
* Tools/WebKitTestRunner/InjectedBundle/gtk/ActivateFontsGtk.cpp:
(WTR::initializeFontConfigSetting):
* Tools/WebKitTestRunner/InjectedBundle/wpe/ActivateFontsWPE.cpp:
(WTR::activateFonts):
* Tools/Scripts/webkitpy/port/gtk_unittest.py

Canonical link: https://commits.webkit.org/256795@main
